File: android/app/src/main/java/app/lociate/android/service/PushRegistrationService.kt
package app.lociate.android.service

import app.lociate.android.data.remote.SupabaseClientProvider
import app.lociate.android.util.SecurePreferences
import com.google.firebase.messaging.FirebaseMessaging
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.postgrest.postgrest
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonPrimitive
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Obtains the FCM registration token and keeps `users.fcm_token` in sync (US-197).
 *
 * Mirrors the iOS `PushRegistrationService`: a token obtained while offline or
 * signed out is queued locally and flushed on the next sign-in or connectivity
 * change, and sign-out clears the column server-side so a shared device stops
 * receiving the previous account's notifications.
 */
@Singleton
class PushRegistrationService @Inject constructor(
    private val supabaseProvider: SupabaseClientProvider,
    private val securePreferences: SecurePreferences
) {
    private val postgrest get() = supabaseProvider.client.postgrest
    private val auth get() = supabaseProvider.client.auth

    /**
     * Requests a token from FCM.
     *
     * Returns without error when Firebase is not configured: a self-hosted build
     * with no google-services.json should still run, just without remote push.
     */
    suspend fun registerForPush() = withContext(Dispatchers.IO) {
        val token = try {
            FirebaseMessaging.getInstance().token.await()
        } catch (e: Exception) {
            // Missing google-services.json, no Play Services, or a network failure.
            // Local geofence notifications are unaffected, so this is not fatal.
            Timber.w(e, "FCM token unavailable — remote push disabled for this session")
            return@withContext
        }

        onTokenReceived(token)
    }

    /** Called from [LociateMessagingService.onNewToken] and after [registerForPush]. */
    suspend fun onTokenReceived(token: String) {
        if (token == securePreferences.getString(KEY_SYNCED_TOKEN)) {
            Timber.d("FCM token unchanged — nothing to upsert")
            return
        }
        securePreferences.putString(KEY_PENDING_TOKEN, token)
        syncPendingToken()
    }

    /**
     * Flushes a queued token. Safe to call repeatedly — no-ops when there is
     * nothing pending or no session.
     */
    suspend fun syncPendingToken() = withContext(Dispatchers.IO) {
        val token = securePreferences.getString(KEY_PENDING_TOKEN) ?: return@withContext
        val authId = auth.currentSessionOrNull()?.user?.id
        if (authId == null) {
            // Signed out: keep it queued for after the next sign-in rather than
            // discarding a token FCM will not hand out again.
            Timber.d("No session — leaving FCM token queued")
            return@withContext
        }

        try {
            postgrest.from("users")
                .update(JsonObject(mapOf("fcm_token" to JsonPrimitive(token)))) {
                    filter { eq("auth_id", authId) }
                }
            securePreferences.putString(KEY_SYNCED_TOKEN, token)
            securePreferences.remove(KEY_PENDING_TOKEN)
            Timber.d("FCM token registered for the current user")
        } catch (e: Exception) {
            // Stays pending; retried on the next sign-in, connectivity change, or
            // app start.
            Timber.e(e, "Failed to upsert FCM token")
        }
    }

    /** Re-binds the current token after a sign-in, since it now belongs to a new account. */
    suspend fun onSignIn() {
        val token = securePreferences.getString(KEY_SYNCED_TOKEN)
            ?: securePreferences.getString(KEY_PENDING_TOKEN)
        if (token != null) {
            securePreferences.putString(KEY_PENDING_TOKEN, token)
            securePreferences.remove(KEY_SYNCED_TOKEN)
        }
        syncPendingToken()
    }

    /**
     * Clears the column server-side, then locally.
     *
     * Order matters: after the session is gone there is no authenticated row to
     * update, and a stale token keeps delivering the previous account's digests to
     * this device.
     */
    suspend fun onSignOut() = withContext(Dispatchers.IO) {
        val authId = auth.currentSessionOrNull()?.user?.id
        if (authId != null) {
            try {
                postgrest.from("users")
                    .update(JsonObject(mapOf("fcm_token" to JsonNull))) {
                        filter { eq("auth_id", authId) }
                    }
                Timber.d("FCM token cleared for the signed-out user")
            } catch (e: Exception) {
                Timber.e(e, "Failed to clear FCM token")
            }
        }
        securePreferences.remove(KEY_SYNCED_TOKEN)
        securePreferences.remove(KEY_PENDING_TOKEN)
    }

    /** After account deletion the row is already gone; drop local state only. */
    fun onAccountDeleted() {
        securePreferences.remove(KEY_SYNCED_TOKEN)
        securePreferences.remove(KEY_PENDING_TOKEN)
    }

    companion object {
        const val KEY_SYNCED_TOKEN = "push_synced_fcm_token"
        const val KEY_PENDING_TOKEN = "push_pending_fcm_token"
    }
}

File: Lociate/Services/PushRegistrationService.swift
import Foundation
import os.log
import Supabase
import UIKit

/// Registers the device with APNs and keeps `users.apns_token` in sync (US-195).
///
/// Everything downstream of a device token already existed — the entitlement
/// declares `aps-environment`, Info.plist declares the `remote-notification`
/// background mode, `users.apns_token` is a column, and the digest edge function
/// reads it — but nothing ever called `registerForRemoteNotifications()`, so the
/// column was always null and every server-side push was skipped.
@Observable
final class PushRegistrationService {
    static let shared = PushRegistrationService()

    private let logger = Logger(subsystem: "app.lociate.ios", category: "PushRegistration")
    private let defaults: UserDefaults
    private let supabase: SupabaseClient

    /// Last token we successfully wrote, so a repeat launch does not re-upsert.
    private static let syncedTokenKey = "push_synced_apns_token"
    /// Token obtained but not yet persisted server-side (offline, or signed out).
    private static let pendingTokenKey = "push_pending_apns_token"

    private(set) var currentToken: String?
    private(set) var lastError: String?

    init(
        defaults: UserDefaults = .standard,
        supabase: SupabaseClient = SupabaseClientProvider.shared
    ) {
        self.defaults = defaults
        self.supabase = supabase
        currentToken = defaults.string(forKey: Self.syncedTokenKey)
            ?? defaults.string(forKey: Self.pendingTokenKey)
    }

    // MARK: - Registration

    /// Asks iOS for a device token.
    ///
    /// Must only be called once notification authorization has been granted:
    /// registering first produces a token the user has not consented to receive
    /// alerts on, and iOS may not deliver one at all.
    @MainActor
    func registerIfAuthorized(isAuthorized: Bool) {
        guard isAuthorized else {
            logger.debug("Notification permission not granted — skipping APNs registration")
            return
        }
        UIApplication.shared.registerForRemoteNotifications()
    }

    // MARK: - Token lifecycle

    /// Called from the app delegate when APNs hands back a device token.
    func handleRegistration(deviceToken: Data) {
        let token = deviceToken.map { String(format: "%02x", $0) }.joined()
        currentToken = token
        lastError = nil

        guard token != defaults.string(forKey: Self.syncedTokenKey) else {
            logger.debug("APNs token unchanged — nothing to upsert")
            return
        }

        defaults.set(token, forKey: Self.pendingTokenKey)
        Task { await syncPendingToken() }
    }

    /// Called from the app delegate when registration fails.
    ///
    /// A failure here must never be fatal: no push is a degraded experience, but
    /// geofence notifications are local and keep working regardless.
    func handleRegistrationFailure(_ error: Error) {
        lastError = error.localizedDescription
        logger.error("APNs registration failed: \(error.localizedDescription, privacy: .public)")
    }

    /// Retries a queued upsert. Safe to call repeatedly — it no-ops when there is
    /// nothing pending, no session, or no connectivity.
    func syncPendingToken() async {
        guard let token = defaults.string(forKey: Self.pendingTokenKey) else { return }

        guard let userId = await currentUserId() else {
            // Signed out: keep the token queued for after the next sign-in.
            logger.debug("No session — leaving APNs token queued")
            return
        }

        do {
            try await supabase
                .from("users")
                .update(["apns_token": token])
                .eq("auth_id", value: userId)
                .execute()

            defaults.set(token, forKey: Self.syncedTokenKey)
            defaults.removeObject(forKey: Self.pendingTokenKey)
            logger.info("APNs token registered for the current user")
        } catch {
            // Stays pending; retried on the next connectivity change, sign-in, or
            // foreground.
            logger.error("Failed to upsert APNs token: \(error.localizedDescription, privacy: .public)")
        }
    }

    /// Called after a successful sign-in, so a token obtained before the user had
    /// a session still reaches the server.
    func handleSignIn() async {
        // Force a re-upsert: this token now belongs to a different account.
        if let token = currentToken {
            defaults.set(token, forKey: Self.pendingTokenKey)
            defaults.removeObject(forKey: Self.syncedTokenKey)
        }
        await syncPendingToken()
    }

    /// Clears the stored token server-side and locally.
    ///
    /// Without this, a signed-out device keeps receiving another account's
    /// digests, and a shared device leaks notifications across users.
    func handleSignOut() async {
        defer {
            defaults.removeObject(forKey: Self.syncedTokenKey)
            defaults.removeObject(forKey: Self.pendingTokenKey)
            currentToken = nil
        }

        guard let userId = await currentUserId() else { return }

        do {
            try await supabase
                .from("users")
                .update(["apns_token": String?.none])
                .eq("auth_id", value: userId)
                .execute()
            logger.info("APNs token cleared for the signed-out user")
        } catch {
            logger.error("Failed to clear APNs token: \(error.localizedDescription, privacy: .public)")
        }
    }

    /// Drops local token state after account deletion. The row is already gone,
    /// so there is nothing to clear server-side.
    func handleAccountDeleted() {
        defaults.removeObject(forKey: Self.syncedTokenKey)
        defaults.removeObject(forKey: Self.pendingTokenKey)
        currentToken = nil
    }

    // MARK: - Helpers

    private func currentUserId() async -> String? {
        do {
            return try await supabase.auth.session.user.id.uuidString
        } catch {
            return nil
        }
    }
}
